sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2AlterTableCommands.scala

```diff
@@ -290,23 +291,45 @@ case class AlterTableCollation(
 }
 
 /**
- * The logical plan of the ALTER TABLE ... ADD CONSTRAINT command.
+ * The logical plan of the ALTER TABLE ... ADD CONSTRAINT command for Primary Key, Foreign Key,
+ * and Unique constraints.
  */
 case class AddConstraint(
     table: LogicalPlan,
     tableConstraint: TableConstraint) extends AlterTableCommand {
+
   override def changes: Seq[TableChange] = {
     val constraint = tableConstraint.toV2Constraint
-    val validatedTableVersion = table match {
-      case t: ResolvedTable if constraint.enforced() =>
-        t.table.currentVersion()
+    // The table version is null because the constraint is not enforced.
+    Seq(TableChange.addConstraint(constraint, null))
+  }
+
+  override protected def withNewChildInternal(newChild: LogicalPlan): LogicalPlan =
+    copy(table = newChild)
+}
+
+/**
+ * The logical plan of the ALTER TABLE ... ADD CONSTRAINT command for Check constraints.
+ * It doesn't extend [[AlterTableCommand]] because its child is a filtered table scan rather than
+ * a table reference.
+ */
+case class AddCheckConstraint(
+    child: LogicalPlan,
+    checkConstraint: CheckConstraint) extends UnaryCommand {
+
+  def changes: Seq[TableChange] = {
+    val constraint = checkConstraint.toV2Constraint
+    val validatedTableVersion = child.find(_.isInstanceOf[DataSourceV2ScanRelation]) match {
+      case Some(d: DataSourceV2ScanRelation) if constraint.enforced() =>
```

Comment:
so users can choose not to enforce the check constraint? do we have a SQL syntax for it?

Reply:
No, it is always enforced. See https://github.com/apache/spark/pull/50772